### ISSUE TITLE
The action 'drop' does not work with mod_http2

### ISSUE THREAD

**Issue by sidney-j-r-m** [SpiderLabs/ModSecurity#992]
The action 'drop' does not work with mod_http2.

**Comment by zimmerle** [comment]
Hi @sidney-j-r-m,

Can you give us more details?

What the debug logs says?

It is only about the "drop" action?

**Comment by sidney-j-r-m** [comment]
Error: Connection drop requested but failed to close the  socket
ModSecurity: Access denied with code 500 (phase 1)

Modules: mod_http2 and mod_security(both compiled from source)
Apache version: 2.4.18-dev(compiled from source)
System: CentOS Linux

Log:
/var/log/httpd/error.****.log:[Sat Nov 28 21:29:32 2015] [error] [client
*.*.*.*] ModSecurity: Access denied with code 500 (phase 1) (Error:
Connection drop requested but failed to close the  socket). Operator EQ
matched 1 at IP:dos_block. [file
"****/modsecurity_crs_11_dos_protection.conf"] [line "*****"] [id "981045"]
[hostname "*****"] [uri "/*****"] [unique_id "VlpG7H8AAAEAAE9fewUAAAAQ"]

On Wed, Dec 9, 2015 at 10:57 AM, Felipe Zimmerle <notifications@github.com>
wrote:

> Hi @sidney-j-r-m <https://github.com/sidney-j-r-m>,
>
> Can you give us more details?
>
> What the debug logs says?
>
> It is only about the "drop" action?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SpiderLabs/ModSecurity/issues/992#issuecomment-163244247>
> .
>



-- 
Sidney

**Comment by sidney-j-r-m** [comment]
Only "drop" action is affected. The "deny" action works fine.

**Comment by zimmerle** [comment]
Hi @sidney-j-r-m,

The "drop" happens to work without the mod_http2 ? Same scenario and rules, but without the mod_http2?

**Comment by sidney-j-r-m** [comment]
"drop" works fine without mod_http2.

**Comment by bazzadp** [comment]
This was discussed on the [mod_http2 project issue list too](https://github.com/icing/mod_h2/issues/127), and seems the fix is as simple as [changing line 269 of mod_security2.c](https://github.com/SpiderLabs/ModSecurity/blob/master/apache2/mod_security2.c#L269) from this:
```
                apr_socket_t *csd = ap_get_module_config(msr->r->connection->conn_config,
                    &core_module);
```
to this:
```
                apr_socket_t *csd;

                /* For mod_http2 used by HTTP/2 there is a virtual connection
                 * so must go through master to get the main connection to drop,
                 * or the drop just doesn't seem to do anything.
                 * For HTTP/1.1 master will not be defined so just go through
                 * normal connection.
                 * More details here: https://github.com/icing/mod_h2/issues/127
                 */
                if (msr->r->connection->master) {
                        csd = ap_get_module_config(msr->r->connection->master->conn_config,
                            &core_module);
                } else {
                        csd = ap_get_module_config(msr->r->connection->conn_config,
                            &core_module);
                }
```

However it's been a looooong time since I did any C coding so will let someone with more experience here judge if that's correctly coded fix, rather than submit a pull request. But seems to work for both HTTP/1.1 and HTTP/2 from my quick tests.

**Comment by sidney-j-r-m** [comment]
Very nice! 
However, I think there is a problem, not just in your code, but also in the 'drop' behavior. If you close an HTTP2 socket connection, for a certain client, all streams of this connection will be closed too, because HTTP2 uses multiplexing. 
But for me, it's okay. The 'drop' action is used to deal with denial of service. In this case, closing the connection and all its streams is the right thing to do.

**Comment by bazzadp** [comment]
Yeah as far as I'm concerned "drop" is a very aggressive, "all bets are off" action. When you use that you really don't care about client handling it nicely at all. And yes I agree it's even more aggressive under HTTP/2 due to multiplexing, whereas some parallel connections for same client might be left alone under HTTP/1.1, but I'd see that as an advantage rather than a disadvantage. In fact would be nice if there was a way to similarly kill ALL connections under HTTP/1.1 :-)

**Comment by sidney-j-r-m** [comment]
I've tested your code in my server and it is working well. Thank you.

**Comment by zimmerle** [comment]
Fixed at #1308.
